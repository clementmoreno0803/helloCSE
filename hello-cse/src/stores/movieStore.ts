import { movieDetail } from "@/models/movieDetailInterface";
import { topMovieInterface } from "@/models/topMovieInterface";
import { upComingMoviesInterface } from "@/models/upCommingMovieInterface";
import { defineStore } from "pinia";
import { computed, ref } from "vue";
import { currentMovieInterface } from "@/models/currentMovieInterface";

export const useMovieStore = defineStore("movieStore", () => {
  const topMovies = ref<topMovieInterface[]>([]);
  const currentMovies = ref<currentMovieInterface[]>([]);
  const upComingMovies = ref<upComingMoviesInterface[]>([]);
  const movieDetails = ref<movieDetail>();
  const inputFilter = ref<string>("");
  const isLoaded = ref<boolean>(false);

  //getter

  const getFilteredMovies = computed(() => {
    return upComingMovies.value.filter((movie) => movie.title.toLowerCase().includes(inputFilter.value.toLowerCase()));
  });

  const getTopFiveMovies = computed(() => {
    const moviesSorted = topMovies.value.sort((a, b) => b.vote_average - a.vote_average);
    return moviesSorted.slice(0, 5);
  });

  const getHeightCurrentMovies = computed(() => {
    return currentMovies.value.slice(0, 8);
  });

  const listTop4Actors = computed(() => {
    const castingSorted = movieDetails.value?.cast.sort((a, b) => b.popularity - a.popularity);
    return castingSorted?.slice(0, 4);
  });

  //setter
  const setTopMovies = (newMovies: topMovieInterface[]) => {
    topMovies.value = newMovies;
  };
  const setCurrentMovies = (newCurrentMovie: currentMovieInterface[]) => {
    currentMovies.value = newCurrentMovie;
  };
  const setUpComingMovies = (newMovies: upComingMoviesInterface[]) => {
    upComingMovies.value = newMovies;
    isLoaded.value = true;
  };
  const setMovieFilterByName = (input: string) => {
    inputFilter.value = input;
  };
  const setMovieDetails = (details: movieDetail | undefined) => {
    if (details) {
      return (movieDetails.value = details);
    }
  };

  return {
    topMovies,
    upComingMovies,
    inputFilter,
    getFilteredMovies,
    getTopFiveMovies,
    getHeightCurrentMovies,
    movieDetails,
    listTop4Actors,
    isLoaded,
    setTopMovies,
    setCurrentMovies,
    setUpComingMovies,
    setMovieFilterByName,
    setMovieDetails
  };
});
