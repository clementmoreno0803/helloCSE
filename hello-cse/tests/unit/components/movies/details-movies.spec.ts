import { describe, it, expect, vi } from "vitest";
import { shallowMount } from '@vue/test-utils';
import detailsMovies from "@/components/movies/details-movies.vue";
import { mockMovieDetail, mockTop4Actors } from "../../fixtures/mockDetailsMovie.mock";
import { createTestingPinia } from "@pinia/testing";
import { createRouter, createWebHistory } from 'vue-router';
import { useMovieStore } from "@/stores/movieStore";

const router = createRouter({
  history: createWebHistory(),
  routes: [
    { path: '/movies/:id', component: detailsMovies },
  ],
});

const generateWrapper = () => {
  return shallowMount(detailsMovies, {
    global: {
      plugins: [
        router,
        createTestingPinia({
          initialState: {
            movieStore: {
              top4Actors: mockTop4Actors,
            },
          },
          createSpy: vi.fn,
        }),
      ],
      mocks: {
        $route: { params: { id: 1 } },
      },
    },
  });
};

describe('detailsMovies.vue', () => {
  it('should render top 4 actors in the cast list',async () => {
    const wrapper = generateWrapper();
    useMovieStore().movieDetails = mockMovieDetail;
    await wrapper.vm.$nextTick()

    const castItems = wrapper.findAll('.movies-details__cast--width');

    expect(castItems.length).toBe(mockMovieDetail.cast.length);
  });
});
