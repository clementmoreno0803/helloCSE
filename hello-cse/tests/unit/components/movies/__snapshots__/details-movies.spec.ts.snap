// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`detailsMovies.vue > should render correctly 1`] = `
<div
  data-v-959aa3f5=""
>
  <v-breadcrumbs
    data-v-959aa3f5=""
    items="[object Object],[object Object]"
  />
  <v-skeleton-loader
    class="movies-details"
    color="rgb(28,28,28)"
    data-v-959aa3f5=""
    elevation="13"
    type="card"
  />
</div>
`;
